apps/admin_panel/assets/src/socket/connector.js

```diff
@@ -0,0 +1,161 @@
+import { appendParams, isAbsoluteURL } from '../utils/url'
+import urlJoin from 'url-join'
+import _ from 'lodash'
+class SocketConnector {
+  constructor (url, params = {}) {
+    this.socket = null
+    this.url = this.normalizeUrl(url)
+    this.params = params
+    this.handleOnConnected = () => {}
+    this.handleOnDisconnected = () => {}
+    this.WebSocket = WebSocket
+    this.connectionStateMap = {
+      '0': 'CONNECTING',
+      '1': 'CONNECTED',
+      '2': 'DISCONNECTING',
+      '3': 'DISCONNECTED'
+    }
+    this.queueJoinChannels = []
+    this.joinedChannels = []
+  }
+  setParams (params) {
+    this.params = params
+  }
+  normalizeUrl (url) {
+    let normalizedUrl = ''
+    if (!isAbsoluteURL(url)) {
+      const origin =
+        window.location.protocol +
+        '//' +
+        window.location.hostname +
+        (window.location.port ? ':' + window.location.port : '')
+      normalizedUrl = urlJoin(origin, url)
+    } else {
+      normalizedUrl = url
+    }
+    return normalizedUrl.replace('http://', 'ws://').replace('https://', 'wss://')
+  }
+  open = (resolve) => () => {
+    console.log('websocket connected.')
+    this.handleOnConnected()
+    this.drainQueue()
+    return resolve(true)
+  }
+  drainQueue = () => {
+    if (this.queueJoinChannels.length > 0) {
+      this.queueJoinChannels.forEach(channel => {
+        this.sendJoinEvent(channel)
+      })
+    }
+  }
+  addJoinedChannelToQueue = () => {
+    if (this.joinedChannels.length > 0) {
+      this.joinedChannels.forEach(channel => {
+        if (!_.includes(this.queueJoinChannels, channel)) {
+          this.queueJoinChannels.push(channel)
+        }
+      })
+      this.joinedChannels = []
+    }
+  }
+  close = () => {
+    this.handleOnDisconnected()
+    this.socket.removeEventListener('open', this.open)
+    this.socket.removeEventListener('close', this.close)
+    this.socket.removeEventListener('message', this.handleMessage)
+    clearInterval(this.heartbeat)
+    console.log('websocket disconnected.')
+    this.addJoinedChannelToQueue()
+    this.reconnect()
+  }
+  on (event, handler) {
+    switch (event) {
+      case 'connected':
+        return (this.handleOnConnected = handler)
+      case 'disconnected':
+        return (this.handleOnDisconnected = handler)
+    }
+  }
+  sendHeartbeatEvent = () => {
+    const payload = JSON.stringify({
+      topic: 'phoenix',
+      event: 'heartbeat',
+      ref: '1',
+      data: {}
+    })
+    this.socket.send(payload)
+  }
+  async reconnect () {
+    setTimeout(async () => {
+      console.log('reconnecting websocket...')
+      await this.connect()
+    }, 5000)
```

Comment:
It is a better idea to extract the timeout constant to another place where is easier to read/update. Plus, it might be reused with another polling mechanism like sending a heartbeat.